File: jasmin-core/django-core/apps/shared/languages.py
"""Single source of truth for the languages the platform ships UI + email
templates for.

Lives in ``apps/shared`` (a foundational, always-shared module) so both the
``accounts`` ``user_language`` field/serializers and the ``notifications``
template registry (``SUPPORTED_LANGUAGES`` / ``DEFAULT_LANGUAGE``) derive from
ONE list — no drift between "what a user can pick" and "what we can render".
"""

from __future__ import annotations

from django.db import models


class LanguageChoices(models.TextChoices):
    EN = "en", "English"
    DE = "de", "German"


# The fallback when a requested language has no on-disk template or DB override.
DEFAULT_LANGUAGE_CODE: str = LanguageChoices.EN.value

# Tuple of bare codes (e.g. ("en", "de")) for membership checks.
SUPPORTED_LANGUAGE_CODES: tuple[str, ...] = tuple(LanguageChoices.values)

File: jasmin-core/react-core/src/shared/i18n/languages.ts
/**
 * Single source of truth for the languages a user can pick and the backend
 * will store.
 *
 * The real list lives in ``apps/shared/languages.py`` (``LanguageChoices``)
 * and reaches the client as the generated ``UserLanguageEnum`` /
 * ``TenantLanguageEnum``. This module adds the two things a schema enum
 * cannot carry — a display label and a flag — so every picker, type guard
 * and payload builder reads one list instead of repeating its own.
 *
 * This is narrower than "languages the UI can render": i18next also ships
 * ``fr`` and ``it`` bundles and will display them if detection selects one.
 * Those are not offered here because the backend's choice fields reject them.
 *
 * Deliberately NOT re-exported from ``./index``: importing that module
 * initialises i18next, and a payload builder that only needs the code list
 * should not boot the whole i18n stack to get it.
 */

import { UserLanguageEnum } from "@shared/api/generated/models";

export interface SupportedLanguage {
  code: UserLanguageEnum;
  /** The language's own name, as a speaker of it would read it. */
  label: string;
  flag: string;
}

/**
 * The order pickers render in, declared rather than derived: German first,
 * because ``de`` is the in-house default and i18next's ``fallbackLng``. The
 * generated enum's own ordering carries no display meaning.
 */
export const SUPPORTED_LANGUAGES: readonly SupportedLanguage[] = [
  { code: UserLanguageEnum.de, label: "Deutsch", flag: "🇩🇪" },
  { code: UserLanguageEnum.en, label: "English", flag: "🇺🇸" },
];

export const SUPPORTED_LANGUAGE_CODES: readonly UserLanguageEnum[] =
  SUPPORTED_LANGUAGES.map((language) => language.code);

/**
 * Narrow an arbitrary locale string to a code the backend accepts.
 *
 * Language detection yields whatever the browser reports — ``fr``, ``it``,
 * ``es``, or a regional tag like ``de-DE`` — while ``user_language`` and
 * ``tenant_language`` are choice fields that 400 on anything outside the
 * list. Any value that came from detection rather than from a picker has to
 * pass through here before it goes into a payload. Regional tags are NOT
 * accepted: split on ``-`` first, so the caller decides what the base of an
 * unsupported regional tag should fall back to.
 */
export function isSupportedLanguageCode(
  value: unknown,
): value is UserLanguageEnum {
  return (
    typeof value === "string" &&
    (SUPPORTED_LANGUAGE_CODES as readonly string[]).includes(value)
  );
}
